src/components/hooks/useProvideAuth.tsx

```diff
@@ -0,0 +1,103 @@
+import React, {
+  createContext,
+  Dispatch,
+  SetStateAction,
+  useContext,
+  useEffect,
+  useState,
+} from 'react';
+
+import { instance } from '../api/api';
+
+export interface IRegisterData {
+  name?: string;
+  email: string;
+  password: string;
+}
+
+export interface IRegisterStatus {
+  error: string | null;
+  loading: boolean;
+  userData: IResponseData | null;
+}
+
+export interface IAuthData {
+  error: string | null;
+  loading: boolean;
+  userData: IResponseData | null;
+  setRegisterData: Dispatch<SetStateAction<IRegisterData | undefined>>;
+}
+
+export interface IResponseData {
+  user: IResponse | null;
+}
+
+export interface IResponse {
+  email: string;
+  bio: string;
+  image: string;
+  token: string;
+}
+
+export const AuthContext = createContext<IAuthData>(null!);
+
+interface IProviderAuth {
+  children: React.ReactNode;
+}
+
+export function ProvideAuth({ children }: IProviderAuth) {
+  const auth = useProvideAuth();
+  return <AuthContext.Provider value={auth}>{children}</AuthContext.Provider>;
+}
+
+export const useAuth = () => {
+  return useContext(AuthContext);
+};
+
+export const useProvideAuth = () => {
+  const [registerData, setRegisterData] = useState<IRegisterData>();
+  const [status, setStatus] = useState<IRegisterStatus>({
+    error: null,
+    userData: null,
+    loading: false,
+  });
+
+  function register(registerData: IRegisterData) {
+    const user = {
+      user: {
+        username: registerData.name,
+        email: registerData.email,
+        password: registerData.password,
+      },
+    };
+    setStatus({ loading: true, userData: null, error: null });
+    instance
+      .post(`users/`, user)
+      .then((res) => setStatus({ userData: res.data, error: null, loading: false }))
+      .catch((error) => setStatus({ error: error, userData: null, loading: false }));
+  }
+
+  function login(loginData: IRegisterData) {
+    const user = {
+      user: {
+        email: loginData.email,
+        password: loginData.password,
+      },
+    };
+    setStatus({ loading: true, userData: null, error: null });
+    instance
+      .post(`users/login`, user)
+      .then((res) => setStatus({ userData: res.data, error: null, loading: false }))
+      .catch((error) => setStatus({ error: error, userData: null, loading: false }));
+  }
+
+  useEffect(() => {
+    if (registerData?.name) {
+      register(registerData);
+    } else if (registerData) {
+      login(registerData);
+    }
+  }, [registerData]);
```

Comment:
I think it would be better to just give a user of the hook access to the login and register functions, like in the example https://tppr.me/eZIVR